io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/wfs/WfsBackedGmlInstanceCollection.java

```diff
@@ -406,12 +448,27 @@ private void proceedOrClose() {
 
 		private void createNextIterator() {
 			URI nextUri;
+			URIBuilder builder;
 			try {
 				nextUri = calculateNextUri();
+				builder = new URIBuilder(nextUri);
 			} catch (URISyntaxException e) {
 				throw new IllegalStateException(e.getMessage(), e);
 			}
 
+			// Download all the files into a folder for Bucket service
+			if (tmpdir != null) {
+				try {
+					URL url = nextUri.toURL();
+					String fileName = tmpdir + url.getPath() + UUID.randomUUID();
```

Comment:
Can the path always be used as file name? I'm sure it for instance may contain `/` and other characters not allowed in a file name.

Related to the UUID - this will mean the file name will be different every time. Did you already think about how to deal with that on the bucket-service side related to if there are already existing files (e.g. avoid duplicate instances being loaded if a dataset is updated automatically from a WFS)

Reply:
In bucket service, I see another problem of how to send these multiple files into a single file as expected [here](https://github.com/wetransform/bucket-service/blob/4ec4b7190eeb472d1cddb95b1c0f43ef0dbda7c1/play/app/wfs/WfsImporter.scala#L160) . 
I cannot write all these different files into a single file. Should I return these several files as a list instead? However, in the end it is still expected as a [seq of Futures of single downloaded file](https://github.com/wetransform/bucket-service/blob/4ec4b7190eeb472d1cddb95b1c0f43ef0dbda7c1/play/app/controllers/Buckets.scala#L1072). If I need to return several files for a single wfs url then I should modify [this code](https://github.com/wetransform/bucket-service/blob/4ec4b7190eeb472d1cddb95b1c0f43ef0dbda7c1/play/app/controllers/Buckets.scala#L1072) as well to handle the scenario where a single wfsEndpoint can have more than one downloaded file. 

Or I was thinking that I can look at the GmlInstanceWriter and StreamGmlWriter to see how it handles the case of pagination and features per request and then replicate the same writing to the files?

Please let me know if you have any further suggestions.